docs/specs/code/TRANSACTION_MEMO.md

```diff
@@ -0,0 +1,87 @@
+     - If user clicks "Yes", the dialog window closes and the textarea become cleared and closes too.
+5. If user have entered the memo and makes a transaction, memo will be saved to localStorage and also to user's external storage.
+6. When user goes to his summary - `/containers/wallet/WalletSummaryPage.js` and clicks on the transaction to expand it, he or she will see his memo at the very bottom of Transaction item (if memo exists).
+7. When user links his external storage account to Yoroi, we check if there are some memos in his external storage. If there are, we download a content of each memo file via Dropbox API, read the content with FileReader object, then we save these memos to localStorage. In this way we can make user be able to install Yoroi on the another device, link his or her external storage account and get user's memos so they can be displayed in WalletSummary.
+
+### Tech spec
+
+#### External storages
+
+We use official [Dropbox Javascript SDK](https://github.com/dropbox/dropbox-sdk-js#readme) for authenticating and making requests to Dropbox API. We can create folders, files, read them and change the content.
+
+To make us able to auth user and manipulate his or her files, we need to [create an app in Dropbox](https://www.dropbox.com/developers/apps/create?_tk=pilot_lp&_ad=topbar5&_camp=create). When we create an app, we can get a `client_id` and we must provide a `redirect_uri` because Dropbox uses Oauth2. Both `client_id` and `redirect_uri` are used to create a link for user, which user should click to authenticate.
+
+Uploading a file: 
+
+When we read a content of Dropbox file, we need to read `Blob` to get the actual content, so we use a `FileReader` object.
+
+#### Storage model
+
+For now we have a separate file for each transaction memo. This approach has it's own pros and cons.
+Pros:
+- It's really easier and requires less code when we know that each file has the only one memo as it's content;
+- There can't be any issues with concurrency when user has multiple instances of the app and both are linked to the same Dropbox account;
+Cons:
+- If user has 10,000 transactions with memos, we'll have to read through Dropbox api 10,000 files, which is bad for application performance.
+
+As for now, the file's name is a transaction id, and the content is the actual memo. In the localStorage memos are saved as an array of objects, where each object looks like this: `{ memoId: someTransactionId, memoText: actualMemoText }`.
+
+#### Concurrency
+
+The border case #1: the user has multiple instances of the app and both are linked to the same Dropbox account
+
+The result of additional researches is: if the user tries to create a new transaction from two instances at the same time, the worst case is Dropbox will create a conflicting copy of the memo file, so we'll have two files with memos, and the last saved version will become a conflicting copy. It's really good for us because we can read both files and merge them to one.
```

Comment:
That means Yoroi would also have to poll existing memos in case they became conflicting and had their contents updated, right? Sounds kind of tricky.

Reply:
Yeah, we need to check if we have a conflict copy actually, because usually we don't. If we have, we gonna check both files, the initial one and the conflict copy, and merge them to one.

Right now this case isn't possible. It will be possible when we switch to one file for all memos, or one file for one wallet.